wa/instrumentation/energy_measurement.py

```diff
@@ -229,30 +257,72 @@ def __init__(self, target, loader=pluginloader, **kwargs):
-                raise ConfigError('No channels found for "{}"'.format(channel))
+            for instrument in self.instruments.itervalues():
+                if not instrument.get_channels(channel):
+                    raise ConfigError('No channels found for "{}"'.format(channel))
 
     def setup(self, context):
-        self.instrumentation.reset(sites=self.sites,
-                                   kinds=self.kinds,
-                                   channels=self.channels)
+        for instrument in self.instruments.itervalues():
+            instrument.reset(sites=self.sites,
+                             kinds=self.kinds,
+                             channels=self.channels)
 
     def start(self, context):
-        self.instrumentation.start()
+        for instrument in self.instruments.itervalues():
+            instrument.start()
 
     def stop(self, context):
-        self.instrumentation.stop()
+        for instrument in self.instruments.itervalues():
+            instrument.stop()
 
     def update_result(self, context):
-        outfile = os.path.join(context.output_directory, 'energy_instrument_output.csv')
-        self.measurement_csv = self.instrumentation.get_data(outfile)
-        context.add_artifact('energy_instrument_output', outfile, 'data')
+        for device, instrument in self.instruments.iteritems():
+            if len(self.instruments) > 1:
+                # If we have multiple instrument instances, we'll need to add a
+                # suffix to the CSV filename so that they don't clash
+                filename = 'energy_instrument_output_{}.csv'.format(device)
+            else:
+                filename = 'energy_instrument_output.csv'
+
+            outfile = os.path.join(context.output_directory, filename)
+            self.measurement_csvs[device] = instrument.get_data(outfile)
+            context.add_artifact('energy_instrument_output', outfile, 'data',
+                                 classifiers={'device': device})
         self.extract_metrics(context)
 
     def extract_metrics(self, context):
-        derived_measurements = DerivedEnergyMeasurements.process(self.measurement_csv)
-        for meas in derived_measurements:
-            context.add_metric(meas.name, meas.value, meas.units)
+        metrics_by_name = defaultdict(list)
+
+        for device in self.instruments:
+            csv = self.measurement_csvs[device]
+            derived_measurements = DerivedEnergyMeasurements.process(csv)
+            for meas in derived_measurements:
+                # If we have multuple instrument instances, add a suffix to the
+                # names of the metrics it reports so that they can be told apart
+                if len(self.instruments) > 1:
+                    metric_name = '{}_{}'.format(meas.name, device)
+                else:
+                    metric_name = meas.name
+                context.add_metric(metric_name, meas.value, meas.units)
```

Comment:
OK - for non-ACME backends the device is `None`, which doesn't look very nice in filenames, so how about the condition becomes `if device is not None`?

Reply:
Erm, sorry I don't understand what you mean. You're already modifying the file names for the artifacts, however their names are always "energy_instrument_output", and you adding the device as the classifier. For metrics you're modifying the metric name. I'm just saying the approaches should be consistent -- either modify the name of the artifact as well as the file names, or do not put device int the metric name and add it as a classifier instead.

Reply:
Ohh, yeah sorry didn't read your comment properly. 

Will have to think for a mo about what happens with metric classifiers when reading the output..

Reply:
If you're scripting WA output processing via the API, the metric/artifact objects will have a `classifiers` attribute. If you're parsing the results.csv, each classifier name will appear as an additional column, depending on the csv processor configuration (it can be configured to show all classifiers or only specific ones).